not compatible with jenkins2

I'm having some issues to configure this plugin from a clean Jenkins 2.10.

I'm getting the following stacktrace

```
javax.servlet.ServletException: java.lang.ClassCastException: org.jenkinsci.plugins.workflow.job.WorkflowJob cannot be cast to hudson.model.AbstractProject
    at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:796)
    at org.kohsuke.stapler.Stapler.invoke(Stapler.java:876)
    at org.kohsuke.stapler.MetaClass$5.doDispatch(MetaClass.java:233)
    at org.kohsuke.stapler.NameBasedDispatcher.dispatch(NameBasedDispatcher.java:58)
    at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:746)
    at org.kohsuke.stapler.Stapler.invoke(Stapler.java:876)
    at org.kohsuke.stapler.Stapler.invoke(Stapler.java:649)
    at org.kohsuke.stapler.Stapler.service(Stapler.java:238)
    at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
    at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:812)
    at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1669)
    at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:135)
    at hudson.util.PluginServletFilter.doFilter(PluginServletFilter.java:126)
    at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
    at hudson.security.csrf.CrumbFilter.doFilter(CrumbFilter.java:49)
    at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
    at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:84)
    at hudson.security.UnwrapSecurityExceptionFilter.doFilter(UnwrapSecurityExceptionFilter.java:51)
    at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
    at jenkins.security.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:117)
    at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
    at org.acegisecurity.providers.anonymous.AnonymousProcessingFilter.doFilter(AnonymousProcessingFilter.java:125)
    at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
    at org.acegisecurity.ui.rememberme.RememberMeProcessingFilter.doFilter(RememberMeProcessingFilter.java:142)
    at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
    at org.acegisecurity.ui.AbstractProcessingFilter.doFilter(AbstractProcessingFilter.java:271)
    at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
    at jenkins.security.BasicHeaderProcessor.doFilter(BasicHeaderProcessor.java:93)
    at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
    at org.acegisecurity.context.HttpSessionContextIntegrationFilter.doFilter(HttpSessionContextIntegrationFilter.java:249)
    at hudson.security.HttpSessionContextIntegrationFilter2.doFilter(HttpSessionContextIntegrationFilter2.java:67)
    at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
    at hudson.security.ChainedServletFilter.doFilter(ChainedServletFilter.java:76)
    at hudson.security.HudsonFilter.doFilter(HudsonFilter.java:171)
    at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
    at org.kohsuke.stapler.compression.CompressionFilter.doFilter(CompressionFilter.java:49)
    at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
    at hudson.util.CharacterEncodingFilter.doFilter(CharacterEncodingFilter.java:82)
    at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
    at org.kohsuke.stapler.DiagnosticThreadNameFilter.doFilter(DiagnosticThreadNameFilter.java:30)
    at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
    at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:585)
    at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
    at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:553)
    at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:223)
    at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1127)
    at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:515)
    at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)
    at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1061)
    at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
    at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
    at org.eclipse.jetty.server.Server.handle(Server.java:499)
    at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:311)
    at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:257)
    at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:544)
    at winstone.BoundedExecutorService$1.run(BoundedExecutorService.java:77)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.ClassCastException: org.jenkinsci.plugins.workflow.job.WorkflowJob cannot be cast to hudson.model.AbstractProject
    at bitbucketpullrequestbuilder.bitbucketpullrequestbuilder.BitbucketBuildTrigger.start(BitbucketBuildTrigger.java:30)
    at org.jenkinsci.plugins.workflow.job.WorkflowJob.submit(WorkflowJob.java:186)
    at hudson.model.Job.doConfigSubmit(Job.java:1229)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:498)
    at org.kohsuke.stapler.Function$InstanceFunction.invoke(Function.java:324)
    at org.kohsuke.stapler.interceptor.RequirePOST$Processor.invoke(RequirePOST.java:52)
    at org.kohsuke.stapler.PreInvokeInterceptedFunction.invoke(PreInvokeInterceptedFunction.java:26)
    at org.kohsuke.stapler.Function.bindAndInvoke(Function.java:167)
    at org.kohsuke.stapler.Function.bindAndInvokeAndServeResponse(Function.java:100)
    at org.kohsuke.stapler.MetaClass$1.doDispatch(MetaClass.java:124)
    at org.kohsuke.stapler.NameBasedDispatcher.dispatch(NameBasedDispatcher.java:58)
    at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:746)
    ... 58 more

```

I hope that someone can help me :)

--- a/pom.xml
+++ b/pom.xml
@@ -3,7 +3,7 @@
   <parent>
     <groupId>org.jenkins-ci.plugins</groupId>
     <artifactId>plugin</artifactId>
-    <version>1.509.4</version><!-- which version of Jenkins is this plugin built against? -->         
+    <version>1.596.1</version>
   </parent>
 
   <artifactId>bitbucket-pullrequest-builder</artifactId>

--- a/src/main/java/bitbucketpullrequestbuilder/bitbucketpullrequestbuilder/BitbucketBuildListener.java
+++ b/src/main/java/bitbucketpullrequestbuilder/bitbucketpullrequestbuilder/BitbucketBuildListener.java
@@ -2,8 +2,12 @@ package bitbucketpullrequestbuilder.bitbucketpullrequestbuilder;
 
 import hudson.Extension;
 import hudson.model.AbstractBuild;
+import hudson.model.Job;
+import hudson.model.Run;
 import hudson.model.TaskListener;
 import hudson.model.listeners.RunListener;
+import hudson.triggers.Trigger;
+import jenkins.model.ParameterizedJobMixIn;
 
 import javax.annotation.Nonnull;
 import java.util.logging.Logger;
@@ -12,25 +16,43 @@ import java.util.logging.Logger;
  * Created by nishio
  */
 @Extension
-public class BitbucketBuildListener extends RunListener<AbstractBuild> {
-    private static final Logger logger = Logger.getLogger(BitbucketBuildTrigger.class.getName());
+public class BitbucketBuildListener extends RunListener<Run<?, ?>> {
+    private static final Logger logger = Logger.getLogger(BitbucketBuildListener.class.getName());
 
     @Override
-    public void onStarted(AbstractBuild abstractBuild, TaskListener listener) {
-        logger.info("BuildListener onStarted called.");
-        BitbucketBuildTrigger trigger = BitbucketBuildTrigger.getTrigger(abstractBuild.getProject());
-        if (trigger == null) {
-            return;
+    public void onStarted(Run r, TaskListener listener) {
+        logger.info("BitbucketBuildListener onStarted called.");
+        BitbucketBuilds builds = builds(r);
+        if (builds != null) {
+            builds.onStarted((BitbucketCause) r.getCause(BitbucketCause.class), r);
         }
-        trigger.getBuilder().getBuilds().onStarted(abstractBuild);
     }
 
     @Override
-    public void onCompleted(AbstractBuild abstractBuild, @Nonnull TaskListener listener) {
-        BitbucketBuildTrigger trigger = BitbucketBuildTrigger.getTrigger(abstractBuild.getProject());
-        if (trigger == null) {
-            return;
+    public void onCompleted(Run r, @Nonnull TaskListener listener) {
+        logger.info("BitbucketBuildListener onCompleted called.");
+        BitbucketBuilds builds = builds(r);
+        if (builds != null) {
+            builds.onCompleted((BitbucketCause) r.getCause(BitbucketCause.class), r.getResult(), r.getUrl());
         }
-        trigger.getBuilder().getBuilds().onCompleted(abstractBuild);
     }
+
+    private BitbucketBuilds builds(Run<?, ?> r) {
+        BitbucketBuildTrigger trigger = null;
+        if (r instanceof AbstractBuild) {
+            trigger = BitbucketBuildTrigger.getTrigger(((AbstractBuild) r).getProject());
+        } else {
+            Job job = r.getParent();
+            if (job instanceof ParameterizedJobMixIn.ParameterizedJob) {
+
+                for (Trigger<?> t : ((ParameterizedJobMixIn.ParameterizedJob) job).getTriggers().values()) {
+                    if (t instanceof BitbucketBuildTrigger) {
+                        trigger = (BitbucketBuildTrigger) t;
+                    }
+                }
+            }
+        }
+        return trigger == null ? null : trigger.getBuilder().getBuilds();
+    }
+
 }

--- a/src/main/java/bitbucketpullrequestbuilder/bitbucketpullrequestbuilder/BitbucketBuildTrigger.java
+++ b/src/main/java/bitbucketpullrequestbuilder/bitbucketpullrequestbuilder/BitbucketBuildTrigger.java
@@ -14,6 +14,7 @@ import hudson.triggers.Trigger;
 import hudson.triggers.TriggerDescriptor;
 import hudson.util.ListBoxModel;
 import jenkins.model.Jenkins;
+import jenkins.model.ParameterizedJobMixIn;
 import net.sf.json.JSONObject;
 import org.apache.commons.lang.StringUtils;
 import org.kohsuke.stapler.DataBoundConstructor;
@@ -33,7 +34,7 @@ import static com.cloudbees.plugins.credentials.CredentialsMatchers.instanceOf;
 /**
  * Created by nishio
  */
-public class BitbucketBuildTrigger extends Trigger<AbstractProject<?, ?>> {
+public class BitbucketBuildTrigger extends Trigger<Job<?, ?>> {
     private static final Logger logger = Logger.getLogger(BitbucketBuildTrigger.class.getName());
     private final String projectPath;
     private final String cron;
@@ -153,7 +154,7 @@ public class BitbucketBuildTrigger extends Trigger<AbstractProject<?, ?>> {
     }
 
     @Override
-    public void start(AbstractProject<?, ?> project, boolean newInstance) {
+    public void start(Job<?, ?> project, boolean newInstance) {
         try {
             this.bitbucketPullRequestsBuilder = BitbucketPullRequestsBuilder.getBuilder();
             this.bitbucketPullRequestsBuilder.setProject(project);
@@ -175,6 +176,16 @@ public class BitbucketBuildTrigger extends Trigger<AbstractProject<?, ?>> {
         return this.bitbucketPullRequestsBuilder;
     }
 
+    private ParameterizedJobMixIn retrieveScheduleJob(final Job<?, ?> job) {
+        // TODO 1.621+ use standard method
+        return new ParameterizedJobMixIn() {
+            @Override
+            protected Job asJob() {
+                return job;
+            }
+        };
+    }
+
     public QueueTaskFuture<?> startJob(BitbucketCause cause) {
         Map<String, ParameterValue> values = this.getDefaultParameters();
 
@@ -183,7 +194,10 @@ public class BitbucketBuildTrigger extends Trigger<AbstractProject<?, ?>> {
             abortRunningJobsThatMatch(cause);
         }
 
-        return this.job.scheduleBuild2(0, cause, new ParametersAction(new ArrayList(values.values())), new RevisionParameterAction(cause.getSourceCommitHash()));
+        return retrieveScheduleJob(this.job).scheduleBuild2(0,
+                new CauseAction(cause),
+                new ParametersAction(new ArrayList(values.values())),
+                new RevisionParameterAction(cause.getSourceCommitHash()));
     }
 
     private void cancelPreviousJobsInQueueThatMatch(@Nonnull BitbucketCause bitbucketCause) {
@@ -239,11 +253,8 @@ public class BitbucketBuildTrigger extends Trigger<AbstractProject<?, ?>> {
 
     @Override
     public void run() {
-        if(this.getBuilder().getProject().isDisabled()) {
-            logger.info("Build Skip.");
-        } else {
-            this.bitbucketPullRequestsBuilder.run();
-        }
+        logger.log(Level.INFO, "TODO: check if project is disabled before calling bitbucketPullRequestsBuilder.run()");
+        this.bitbucketPullRequestsBuilder.run();
         this.getDescriptor().save();
     }
 
@@ -259,7 +270,7 @@ public class BitbucketBuildTrigger extends Trigger<AbstractProject<?, ?>> {
 
         @Override
         public boolean isApplicable(Item item) {
-            return true;
+            return item instanceof Job && item instanceof ParameterizedJobMixIn.ParameterizedJob;
         }
 
         @Override

--- a/src/main/java/bitbucketpullrequestbuilder/bitbucketpullrequestbuilder/BitbucketBuildTrigger.java
+++ b/src/main/java/bitbucketpullrequestbuilder/bitbucketpullrequestbuilder/BitbucketBuildTrigger.java
@@ -253,9 +253,13 @@ public class BitbucketBuildTrigger extends Trigger<Job<?, ?>> {
 
     @Override
     public void run() {
-        logger.log(Level.INFO, "TODO: check if project is disabled before calling bitbucketPullRequestsBuilder.run()");
-        this.bitbucketPullRequestsBuilder.run();
-        this.getDescriptor().save();
+    	Job<?,?> project = this.getBuilder().getProject();
+    	if (project instanceof AbstractProject && ((AbstractProject)project).isDisabled()) {
+    		logger.info("Build Skip.");
+    	} else {
+    		this.bitbucketPullRequestsBuilder.run();
+            this.getDescriptor().save();
+    	}
     }
 
     @Override

--- a/src/main/java/bitbucketpullrequestbuilder/bitbucketpullrequestbuilder/BitbucketBuilds.java
+++ b/src/main/java/bitbucketpullrequestbuilder/bitbucketpullrequestbuilder/BitbucketBuilds.java
@@ -1,10 +1,7 @@
 package bitbucketpullrequestbuilder.bitbucketpullrequestbuilder;
 
 import bitbucketpullrequestbuilder.bitbucketpullrequestbuilder.bitbucket.BuildState;
-import hudson.model.AbstractBuild;
-import hudson.model.Cause;
-import hudson.model.Result;
-import jenkins.model.Jenkins;
+import hudson.model.*;
 import jenkins.model.JenkinsLocationConfiguration;
 
 import java.io.IOException;
@@ -24,16 +21,7 @@ public class BitbucketBuilds {
         this.repository = repository;
     }
 
-    public BitbucketCause getCause(AbstractBuild build) {
-        Cause cause = build.getCause(BitbucketCause.class);
-        if (cause == null || !(cause instanceof BitbucketCause)) {
-            return null;
-        }
-        return (BitbucketCause) cause;
-    }
-
-    public void onStarted(AbstractBuild build) {
-        BitbucketCause cause = this.getCause(build);
+    void onStarted(BitbucketCause cause, Run<?, ?> build) {
         if (cause == null) {
             return;
         }
@@ -44,24 +32,21 @@ public class BitbucketBuilds {
         }
     }
 
-    public void onCompleted(AbstractBuild build) {
-        BitbucketCause cause = this.getCause(build);
+    void onCompleted(BitbucketCause cause, Result result, String buildUrl) {
         if (cause == null) {
             return;
         }
-        Result result = build.getResult();
         JenkinsLocationConfiguration globalConfig = new JenkinsLocationConfiguration();
         String rootUrl = globalConfig.getUrl();
-        String buildUrl = "";
         if (rootUrl == null) {
             logger.warning("PLEASE SET JENKINS ROOT URL IN GLOBAL CONFIGURATION FOR BUILD STATE REPORTING");
         } else {
-            buildUrl = rootUrl + build.getUrl();
+            buildUrl = rootUrl + buildUrl;
             BuildState state = result == Result.SUCCESS ? BuildState.SUCCESSFUL : BuildState.FAILED;
             repository.setBuildStatus(cause, state, buildUrl);
         }
 
-        if ( this.trigger.getApproveIfSuccess() && result == Result.SUCCESS ) {
+        if (this.trigger.getApproveIfSuccess() && result == Result.SUCCESS) {
             this.repository.postPullRequestApproval(cause.getPullRequestId());
         }
     }

--- a/src/main/java/bitbucketpullrequestbuilder/bitbucketpullrequestbuilder/BitbucketPullRequestsBuilder.java
+++ b/src/main/java/bitbucketpullrequestbuilder/bitbucketpullrequestbuilder/BitbucketPullRequestsBuilder.java
@@ -1,7 +1,6 @@
 package bitbucketpullrequestbuilder.bitbucketpullrequestbuilder;
 
 import bitbucketpullrequestbuilder.bitbucketpullrequestbuilder.bitbucket.Pullrequest;
-import hudson.model.AbstractProject;
 
 import java.io.UnsupportedEncodingException;
 import java.security.MessageDigest;
@@ -11,6 +10,7 @@ import java.util.Collection;
 import java.util.logging.Level;
 import java.util.logging.Logger;
 
+import hudson.model.Job;
 import org.apache.commons.codec.binary.Hex;
 
 /**
@@ -18,7 +18,7 @@ import org.apache.commons.codec.binary.Hex;
  */
 public class BitbucketPullRequestsBuilder {
     private static final Logger logger = Logger.getLogger(BitbucketBuildTrigger.class.getName());
-    private AbstractProject<?, ?> project;
+    private Job<?, ?> project;
     private BitbucketBuildTrigger trigger;
     private BitbucketRepository repository;
     private BitbucketBuilds builds;
@@ -47,7 +47,7 @@ public class BitbucketPullRequestsBuilder {
         return this;
     }
 
-    public void setProject(AbstractProject<?, ?> project) {
+    public void setProject(Job<?, ?> project) {
         this.project = project;
     }
 
@@ -55,7 +55,7 @@ public class BitbucketPullRequestsBuilder {
         this.trigger = trigger;
     }
 
-    public AbstractProject<?, ?> getProject() {
+    public Job<?, ?> getProject() {
         return this.project;
     }        
     
@@ -73,6 +73,7 @@ public class BitbucketPullRequestsBuilder {
         logger.log(Level.WARNING, "Failed to produce hash", exc);
       }
       return this.project.getFullName();
+
     }
 
     public BitbucketBuildTrigger getTrigger() {

